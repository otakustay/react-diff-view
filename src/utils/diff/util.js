export const first = array => array[0];

export const last = array => array[array.length - 1];

export const sideToProperty = side => [side + 'Start', side + 'Lines'];
