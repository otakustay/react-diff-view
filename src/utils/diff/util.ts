export const first = <T>(array: T[]): T => array[0];

export const last = <T>(array: T[]): T => array[array.length - 1];

export const sideToProperty = (side: 'old'| 'new') => [side + 'Start', side + 'Lines'];
