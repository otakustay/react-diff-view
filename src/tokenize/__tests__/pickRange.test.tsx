// @ts-nocheck ignore: third-party
import * as renderer from 'react-test-renderer';
import {Diff, Hunk, parseDiff, tokenize, markEdits, pickRanges} from '../..';

describe('pickRange', () => {
    test('for definition', () => {
        const oldSource = `// Copyright

package com.xxx;

import java.util.List;
import java.util.Set;`;

        const diffText = `diff --git a/a b/b
index 5b25e5b..772d084 100644
--- a/a
+++ b/b
@@ -2,6 +2,7 @@
 
 package com.xxx;
 
+import java.util.Date;
 import java.util.List;
 import java.util.Set;`;

        const defs = [{
            id: 'yz5k9m0BvISpUQ2asedw',
            type: 'def',
            row: 3,
            col: 9,
            binding: 'package',
            length: 7,
            token: 'com.xxx',
        }];

        const renderToken = (token, defaultRender, i) => {
            switch (token.type) {
                case 'ref':
                case 'def':
                    return <span key={`def-${i}`} className="def">{defaultRender(token.children[0], i)}</span>;
                case 'space':
                default:
                    return defaultRender(token, i);
            }
        };

        const tokenizer = hunks => {
            if (!hunks) {
                return null;
            }

            const options = {
                highlight: false,
                enhancers: [
                    markEdits(hunks, {type: 'block'}),
                    pickRanges([], defs.map(i => ({...i, lineNumber: i.row, start: i.col - 1}))),
                ],
            };

            try {
                return tokenize(hunks, options);
            } catch (ex) {
                return null;
            }
        };

        const [diff] = parseDiff(diffText, {nearbySequences: 'zip'});

        const {type, hunks} = diff;

        const tokens = tokenizer(hunks);

        /* eslint-disable react/jsx-no-bind */
        expect(renderer.create(
            <Diff
                viewType="split"
                diffType={type}
                hunks={hunks}
                tokens={tokens}
                renderToken={renderToken}
                oldSource={oldSource}
            >
                {hunks => hunks.map(hunk => <Hunk key={hunk.content} hunk={hunk} />)}
            </Diff>
        ).toJSON()).toMatchSnapshot();
    });
});
