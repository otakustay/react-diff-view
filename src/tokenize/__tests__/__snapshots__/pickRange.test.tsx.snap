// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pickRange for definition 1`] = `
<table
  className="diff diff-split"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <tbody
    className="diff-hunk"
  >
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        2
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        2
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </td>
    </tr>
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        3
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        package com.xxx;
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        3
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        package 
        <span
          className="def"
        >
          com.xxx
        </span>
        ;
      </td>
    </tr>
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        4
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        4
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        
      </td>
    </tr>
    <tr
      className="diff-line diff-line-new-only"
    >
      <td
        className="diff-gutter diff-gutter-omit"
      />
      <td
        className="diff-code diff-code-omit"
      />
      <td
        className="diff-gutter diff-gutter-insert"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        5
      </td>
      <td
        className="diff-code diff-code-insert"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        import java.util.Date;
      </td>
    </tr>
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        5
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        import java.util.List;
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        6
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        import java.util.List;
      </td>
    </tr>
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        6
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        import java.util.Set;
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        7
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        import java.util.Set;
      </td>
    </tr>
  </tbody>
</table>
`;
